### PR TITLE
nixos/doc/rl-2211: document nsncd option

### PR DESCRIPTION
**Pull request by flokli**
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [x] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 1]
```diff
@@ -38,6 +38,17 @@ In addition to numerous new and upgraded packages, this release has the followin
...
+- `nscd` functionality, necessary to provide non-glibc-builtin NSS
+  modules (such as `nss_systemd` or `nss_ldap`) can now be provided by
+  `nsncd`, by setting `services.nscd.enableNsncd` to `true`.
+
+  The glibc-provided `nscd` daemon wasn't very reliable, and it was not fully
+  possible to disable caching functionality, causing various issues and stuck
+  lookups.
+
+  The plan is to switch from `glibc-nscd` to `nsncd` by default for the release
+  after this.
```
Comment (winterqt) [inline nixos/doc/manual/release-notes/rl-2211.section.md:50]: These subpoints should be prefixed with `-` (I think).

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 2]
```diff
@@ -38,6 +38,17 @@ In addition to numerous new and upgraded packages, this release has the followin
   for a transition period so that in time the ecosystem can switch without
   breaking compatibility with any supported NixOS release.
 
+- `nscd` functionality, necessary to provide non-glibc-builtin NSS
+  modules (such as `nss_systemd` or `nss_ldap`) can now be provided by
+  `nsncd`, by setting `services.nscd.enableNsncd` to `true`.
+
+  The glibc-provided `nscd` daemon wasn't very reliable, and it was not fully
+  possible to disable caching functionality, causing various issues and stuck
+  lookups.
```
Comment (winterqt) [inline nixos/doc/manual/release-notes/rl-2211.section.md:47]:
```suggestion
  The `nscd` daemon provided by glibc isn't very reliable, and it's not fully
  possible to disable caching functionality, causing various issues and failed
  lookups.
```

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 1 above)
Comment (winterqt) [inline nixos/doc/manual/release-notes/rl-2211.section.md:50]:
```suggestion
  We plan to use `nsncd` by default in NixOS 23.05.
```

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 3]
```diff
@@ -38,6 +38,17 @@ In addition to numerous new and upgraded packages, this release has the followin
   for a transition period so that in time the ecosystem can switch without
   breaking compatibility with any supported NixOS release.
 
+- `nscd` functionality, necessary to provide non-glibc-builtin NSS
+  modules (such as `nss_systemd` or `nss_ldap`) can now be provided by
+  `nsncd`, by setting `services.nscd.enableNsncd` to `true`.
```
Comment (winterqt) [inline nixos/doc/manual/release-notes/rl-2211.section.md:43]: Maybe link nsncd's home page?

**Review by winterqt — COMMENTED** [review]
Might be missing the mark here, sorry in advance if these comments aren't too good.

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 1 above)
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-2211.section.md:50]: Makes sense. Updated.

Comment (flokli) [reply nixos/doc/manual/release-notes/rl-2211.section.md:50]: Suggestion applied, thanks!

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 2 above)
Comment (flokli) [reply nixos/doc/manual/release-notes/rl-2211.section.md:47]: Suggestion applied, thanks!

**Comment by winterqt** [comment]
I think the main issue I want to address with this note is *why* we're switching to this alternate implementation. I know you left a note about glibc's impl not being able to disable caching, but like... I don't know.

I want to give you a clearer suggestion here, but I would like some more info first.

Why did we decide to switch? Are other distributions using this implementation? If not, what justification do we have for being on the bleeding edge by eventually making this the default?

I think that context would help me figure out what to say and how to say it.

**Comment by flokli** [comment]
> I think the main issue I want to address with this note is _why_ we're switching to this alternate implementation. I know you left a note about glibc's impl not being able to disable caching, but like... I don't know.
> 
> I want to give you a clearer suggestion here, but I would like some more info first.
> 
> Why did we decide to switch? Are other distributions using this implementation? If not, what justification do we have for being on the bleeding edge by eventually making this the default?
> 
> I think that context would help me figure out what to say and how to say it.

https://github.com/NixOS/nixpkgs/issues/135888

**Comment by winterqt** [comment]
Admittedly, I don't know a lot about nss. But, just so I understand things:

> But when I run any application that resolves the same DNS name in a loop on a clean NixOS system, I observe that DNS packets are not sent on each request, they are only sent after ttl elapses.

Is this not something we want by default, for whatever reason? Why would we want to disobey the TTL (unless the cache is flushed)?

Sorry if this is a really bad question or something 😅

**Comment by flokli** [comment]
The `nscd` file socket is about /dispatching/ the NSS request to a daemon on the system, which takes care of `dlopen()`ing a working NSS module into its own memory.

glibc-nscd also had a generic not-very-smart cache for the requests it previously replied to, which caused various problems. `nsncd` avoids these by not doing any caching on its own on that layer, and really just acting as a NSS lookup dispatcher.

Regular distros usually don't run nscd or nsncd so binaries load NSS modules directly. I assume these NSS modules still have their own caches, and honor DNS TTL there.

With nsncd in the mix, this inside-each-nss-module kind of caching would still be present.

**Comment by flokli** [comment]
@winterqt I described the whole situation about why glibc-nscd is problematic over in https://flokli.de/posts/2022-11-18-nsncd/. Let me know what you think of this should still end up in the release notes.

The main intent for the release notes is making people aware of the option to test it, so we can switch the default eventually.

**Comment by flokli** [comment]
I incorporated some of the proposed rewordings.

**Pull request merged**